### PR TITLE
Prevents DS regexes with non-consecutive order from generating invalid CRconfig/Snapshot

### LINKED ISSUE #3548: TR: DS regexes with non-consecutive order generate invalid CRconfig/Snapshot
By ignorance or by maybe by design, I created a CR-config that got rejected from Traffic router. I added a `HOST_REGEXP` with order of "10" which shouldn't have been a problem (there is no default value). 

This created a bunch of null entries added in the CR-Config which Traffic router was unable to parse.

Note: CR-Config rejections are not being relayed back to users in Portal.

```
  "test-ds": {
    "coverageZoneOnly": "false",
    "deepCachingType": "NEVER",
    "domains": [
      "test-ds.wcdn.cox.net"
    ],
    "geolocationProvider": "maxmindGeolocationService",
    "ip6RoutingEnabled": "false",
    "matchsets": [
      {
        "matchlist": [
          {
            "match-type": "HOST",
            "regex": ".*\\.test-ds\\..*"
          }
        ],
        "protocol": "DNS"
      },
++    null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      {
        "matchlist": [
          {
            "match-type": "HOST",
            "regex": "mycname.cdn.cox.net"
          }
        ],
        "protocol": "DNS"
      }
    ],
```

Traffic Router logs:

```
WARN  2019-05-06T12:51:56.861 [New I/O worker #2] com.comcast.cdn.traffic_control.traffic_router.core.monitor.TrafficMonitorWatcher - com.comcast.cdn.traffic_control.traffic_router.core.util.JsonUtilsException: Failed querying JSON for key: protocol
com.comcast.cdn.traffic_control.traffic_router.core.util.JsonUtilsException: Failed querying JSON for key: protocol
        at com.comcast.cdn.traffic_control.traffic_router.core.util.JsonUtils.throwException(JsonUtils.java:131)
        at com.comcast.cdn.traffic_control.traffic_router.core.util.JsonUtils.getString(JsonUtils.java:104)
        at com.comcast.cdn.traffic_control.traffic_router.core.config.ConfigHandler.parseDeliveryServiceConfig(ConfigHandler.java:439)
        at com.comcast.cdn.traffic_control.traffic_router.core.config.ConfigHandler.processConfig(ConfigHandler.java:176)
        at com.comcast.cdn.traffic_control.traffic_router.core.monitor.TrafficMonitorWatcher$2.update(TrafficMonitorWatcher.java:116)
        at com.comcast.cdn.traffic_control.traffic_router.core.util.PeriodicResourceUpdater.updateDatabase(PeriodicResourceUpdater.java:165)
        at com.comcast.cdn.traffic_control.traffic_router.core.util.PeriodicResourceUpdater$UpdateHandler.onCompleted(PeriodicResourceUpdater.java:285)
        at com.comcast.cdn.traffic_control.traffic_router.core.util.PeriodicResourceUpdater$UpdateHandler.onCompleted(PeriodicResourceUpdater.java:256)
        at com.ning.http.client.AsyncCompletionHandler.onCompleted(AsyncCompletionHandler.java:63)
        at com.ning.http.client.providers.netty.NettyResponseFuture.getContent(NettyResponseFuture.java:279)
        at com.ning.http.client.providers.netty.NettyResponseFuture.done(NettyResponseFuture.java:307)
        at com.ning.http.client.providers.netty.NettyAsyncHttpProvider.markAsDone(NettyAsyncHttpProvider.java:1454)
        at com.ning.http.client.providers.netty.NettyAsyncHttpProvider.finishUpdate(NettyAsyncHttpProvider.java:1470)
        at com.ning.http.client.providers.netty.NettyAsyncHttpProvider.access$2600(NettyAsyncHttpProvider.java:158)
        at com.ning.http.client.providers.netty.NettyAsyncHttpProvider$HttpProtocol.handle(NettyAsyncHttpProvider.java:2229)
        at com.ning.http.client.providers.netty.NettyAsyncHttpProvider.messageReceived(NettyAsyncHttpProvider.java:1163)
        at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
        at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
        at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
        at org.jboss.netty.handler.stream.ChunkedWriteHandler.handleUpstream(ChunkedWriteHandler.java:142)
        at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
        at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
        at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
        at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:459)
        at org.jboss.netty.handler.codec.replay.ReplayingDecoder.callDecode(ReplayingDecoder.java:536)
        at org.jboss.netty.handler.codec.replay.ReplayingDecoder.messageReceived(ReplayingDecoder.java:435)
        at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
        at org.jboss.netty.handler.codec.http.HttpClientCodec.handleUpstream(HttpClientCodec.java:92)
        at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
        at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
        at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
        at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
        at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
        at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:109)
        at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:312)
        at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:90)
        at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
        at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
        at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

### PR DESCRIPTION
**Pull request by srijeet0406**
## What does this PR (Pull Request) do?

- [x] This PR fixes #3548 

## Which Traffic Control components are affected by this PR?

- CDN in a Box
- Documentation
- Traffic Ops
- Traffic Router

## What is the best way to verify this PR?

Run CDN-in-a-box.
For any CDN, go into the associated delivery services.
In one of the delivery services, go to the regexes. Try adding another regex with the same order as one that exists.
You should get an error saying that you cannot add two regexes with the same order.
Now try adding a regex with a different order, which is not the consecutive number of whats already present. For ex, if you have a regex of order 0, try adding another regex of order 20 (or something thats not 1).
Do a snapshot of the CDN associated with this DS now.
If you do a `GET` for the snapshot of this CDN, you should not see `null` values in the delivery service matchlists. Whatever order is not present, should just not show up in the CRConfig file.

## If this is a bug fix, what versions of Traffic Control are affected?

- master

## The following criteria are ALL met by this PR

- [x] This PR includes tests
- [x] This PR includes documentation
- [x] This PR includes an update to CHANGELOG.md 
- [x] This PR includes any and all required license headers
- [x] This PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

**Comment by mitchell852** [comment]
@srijeet0406 - think you can rename this to something like:

`Prevents DS regexes with non-consecutive order from generating invalid CRconfig"

or something like that?

File: traffic_ops/traffic_ops_golang/deliveryservicesregexes/deliveryservicesregexes.go [hunk 1]
```diff
@@ -332,6 +342,18 @@ func validateDSRegexType(tx *sql.Tx, typeID int) error {
 	return err
 }
 
+func validateDSRegexOrder(tx *sql.Tx, dsID int, order int) error {
```
Comment (mitchell852) [inline traffic_ops/traffic_ops_golang/deliveryservicesregexes/deliveryservicesregexes.go:345]: looks like i can create a regex with order < 0. want to fix that?

Comment (srijeet0406) [reply traffic_ops/traffic_ops_golang/deliveryservicesregexes/deliveryservicesregexes.go:345]: Fixed now. TP stops it by default, so I missed changing it in the API.

**Comment by ocket8888** [comment]
This doesn't appear to actually touch documentation.

File: CHANGELOG.md [hunk 2]
```diff
@@ -54,6 +54,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed ORT bug miscalculating Mid Max Origin Connections as all servers, usually resulting in 1.
 - Added Delivery Service Raw Remap `__RANGE_DIRECTIVE__` directive to allow inserting the Range Directive after the Raw Remap text. This allows Raw Remaps which manipulate the Range.
 - Added an option for `coordinateRange` in the RGB configuration file, so that in case a client doesn't have a postal code, we can still determine if it should be allowed or not, based on whether or not the latitude/ longitude of the client falls within the supplied ranges. [Related github issue](https://github.com/apache/trafficcontrol/issues/4372)
+- Fixed #3548 - Setting order in Portal can create failing CR-config
```
Comment (mitchell852) [inline CHANGELOG.md:57]: Can you make this a little more descriptive. maybe just use the PR title? `Prevents DS regexes with non-consecutive order from generating invalid CRconfig`

File: traffic_ops/traffic_ops_golang/crconfig/deliveryservice.go [hunk 3]
```diff
@@ -491,12 +492,14 @@ order by dr.set_number asc
 		if err := rows.Scan(&pattern, &ttype, &dstype, &setnum, &dsname); err != nil {
 			return nil, nil, errors.New("scanning deliveryservice regexes: " + err.Error())
 		}
-
+		if _, ok := dsNameOrderMap[dsname]; !ok {
```
Comment (mitchell852) [inline traffic_ops/traffic_ops_golang/crconfig/deliveryservice.go:495]: do you think you can add a comment here to describe what you're doing (removing gaps)

**Comment by mitchell852** [comment]
@srijeet0406 - looks like the Go Unit Tests check is failing

**Comment by srijeet0406** [comment]
@mitchell852 funny, I added a period at the end of the comment and it passes again :)

**Pull request merged**